### PR TITLE
[Merged by Bors] - feat: better handling of symm in library_search

### PR DESCRIPTION
**Pull request by kim-em**
Rather than separately indexing the `symm` version of each `Eq` lemma (and `Iff` lemma, although we were forgetting to do this at all!), we just try `library_search` on the goal and then on `symm` of the goal.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: test/LibrarySearch/basic.lean [hunk 1]
```diff
@@ -135,3 +136,10 @@ assert_no_sorry Bool_eq_iff2
 
 -- Example from https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/library_search.20regression/near/354025788
 example {r : α → α → Prop} : Function.Surjective (Quot.mk r) := by library_search
+
+set_option trace.Tactic.librarySearch true
```
Comment (alexjbest) [inline test/LibrarySearch/basic.lean:140]: Do you want these lines in the test code still?

File: Mathlib/Tactic/LibrarySearch.lean [hunk 2]
```diff
@@ -37,18 +37,18 @@ initialize registerTraceClass `Tactic.librarySearch.lemmas
 /--
 A "modifier" for a declaration.
 * `none` indicates the original declaration,
-* `symm` indicates that (possibly after binders) the declaration is an `=`,
+* `symm` indicates that (possibly after binders) the declaration is an `=` or `↔`,
```
Comment (alexjbest) [inline Mathlib/Tactic/LibrarySearch.lean:40]: This modifier is gone now

File: Mathlib/Tactic/LibrarySearch.lean [hunk 3]
```diff
@@ -37,18 +37,18 @@ initialize registerTraceClass `Tactic.librarySearch.lemmas
...
   and we want to consider the symmetric version,
```
Comment (semorrison) [inline Mathlib/Tactic/LibrarySearch.lean:41]:
```suggestion
```

File: Mathlib/Tactic/LibrarySearch.lean [hunk 4]
```diff
@@ -163,6 +158,18 @@ def librarySearchCore (goal : MVarId)
...
+-/
+def librarySearchSymm (goal : MVarId)
+    (required : List Expr) (solveByElimDepth := 6) :
+    ListM MetaM (MetavarContext × List MVarId) :=
+  .append (librarySearchCore goal required solveByElimDepth) <| .squash do
+    if let some symm ← try? goal.symm then
+      return librarySearchCore symm required solveByElimDepth
+    else
+      return .nil
+
```
[the remark this answers is not in the document]
Comment (semorrison) [reply Mathlib/Tactic/LibrarySearch.lean:170]: solveByElim will still use local hypotheses in both their forward and backwards versions, where relevant. This is only changing how we handle the single library lemma which is applied first.

[the remark this answers is not in the document]
Comment (kmill) [reply Mathlib/Tactic/LibrarySearch.lean:170]: Ah, I see, I forgot that library search never composes more than one library lemma.

**Comment by kmill** [comment]
Thanks!

bors r+

**Pull request closed without merging**